trieste/models/gpflow/models.py

```diff
@@ -1228,13 +1228,19 @@ def _ensure_variable_model_data(self) -> None:
                 ),
             )
 
+            # reinitialise the model so that the underlying Parameters have the right shape
+            # and then reassign the original values
+            old_q_mu = model.q_mu
+            old_q_sqrt = model.q_sqrt
             model.__init__(  # type: ignore[misc]
                 variable_data,
                 model.kernel,
                 model.likelihood,
                 model.mean_function,
                 model.num_latent_gps,
             )
+            model.q_mu.assign(old_q_mu)
+            model.q_sqrt.assign(old_q_sqrt)
```

Comment:
The VGP trieste model follows a different pattern to the others, i.e. it recreates the gpflow model; which seems messy. Also, update to `model.num_data` is missing. Would it be better to follow, e.g., what SGPR does?

```suggestion
        if not all(is_variable(x) and x.shape[0] is None for x in self.model.data):
            self.model.data = (
                tf.Variable(
                    self._model.data[0],
                    trainable=False,
                    shape=[None, *self._model.data[0].shape[1:]],
                ),
                tf.Variable(
                    self._model.data[1],
                    trainable=False,
                    shape=[None, *self._model.data[1].shape[1:]],
                ),
            )

        if not is_variable(self.model.num_data):
            self.model.num_data = tf.Variable(self.model.num_data, trainable=False)
```

Reply:
The problem is that VGP uses more complex internal state, including a couple of GPflow Parameters for q_mu and q_sqrt that also need to updated to be dynamically shaped. We could copy more of the VGP initialiser into this method, but that actually makes the code more complex not less. (Regarding num_data, the initializer converts that to a dynamically shaped Parameter too.)

Reply:
And as far as I know there is no simpler way of changing the shape of gpflow Parameters than recreating them and assigning the old value.